drivers/crc/crc_sf32lb.c

```diff
@@ -0,0 +1,288 @@
+{
+	struct crc_sf32lb_data *data = dev->data;
+
+	k_sem_give(&data->lock);
+}
+
+static uint32_t crc_sf32lb_get_result(const struct device *dev)
+{
+	const struct crc_sf32lb_config *config = dev->config;
+	struct crc_sf32lb_data *data = dev->data;
+	uint32_t raw;
+	uint32_t mask;
+
+	raw = sys_read32(config->base + CRC_DR_OFFSET) ^ data->xor_out;
+	mask = crc_sf32lb_mask(data->width);
+
+	return raw & mask;
+}
+
+static int crc_sf32lb_begin(const struct device *dev, struct crc_ctx *ctx)
+{
+	const struct crc_sf32lb_config *config = dev->config;
+	struct crc_sf32lb_data *data = dev->data;
+	uint8_t polysize;
+	uint8_t width;
+	uint32_t xor_out;
+	uint32_t cr;
+	uint32_t mask;
+	int ret;
+
+	if ((ctx == NULL) || (ctx->state != CRC_STATE_IDLE)) {
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	ret = crc_sf32lb_prepare_config(ctx, &polysize, &width, &xor_out);
+	if (ret != 0) {
+		crc_sf32lb_unlock(dev);
+		return ret;
+	}
+
+	data->width = width;
+	data->xor_out = xor_out;
+	mask = crc_sf32lb_mask(width);
+
+	cr = FIELD_PREP(CRC_CR_DATASIZE_Msk, CRC_DATASIZE_8) |
+	     FIELD_PREP(CRC_CR_POLYSIZE_Msk, polysize);
+
+	if ((ctx->reversed & CRC_FLAG_REVERSE_INPUT) != 0U) {
+		cr |= FIELD_PREP(CRC_CR_REV_IN_Msk, CRC_REV_IN_BYTE);
+	}
+
+	if ((ctx->reversed & CRC_FLAG_REVERSE_OUTPUT) != 0U) {
+		cr |= CRC_CR_REV_OUT;
+	}
+
+	sys_write32(cr, config->base + CRC_CR_OFFSET);
+	sys_write32(ctx->seed & mask, config->base + CRC_INIT_OFFSET);
+	sys_write32(ctx->polynomial & mask, config->base + CRC_POL_OFFSET);
+
+	/* Reset data register to the provided seed */
+	sys_write32(cr | CRC_CR_RESET, config->base + CRC_CR_OFFSET);
+	sys_write32(cr, config->base + CRC_CR_OFFSET);
+
+	ctx->state = CRC_STATE_IN_PROGRESS;
+	ctx->result = ctx->seed & mask;
+
+	return 0;
+}
+
+static int crc_sf32lb_update(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			     size_t bufsize)
+{
+	const struct crc_sf32lb_config *config = dev->config;
+	const uint8_t *data_buf = buffer;
+
+	if (ctx->state != CRC_STATE_IN_PROGRESS) {
+		return -EINVAL;
+	}
+
+	if ((bufsize > 0U) && (data_buf == NULL)) {
+		ctx->state = CRC_STATE_IDLE;
+		crc_sf32lb_unlock(dev);
+		return -EINVAL;
+	}
+
+	size_t aligned_len = ROUND_DOWN(bufsize, sizeof(uint32_t));
+	size_t idx = 0U;
+
+	for (; idx < aligned_len; idx += sizeof(uint32_t)) {
+		uint32_t data = sys_get_le32(&data_buf[idx]);
+
+		sys_write32(data, config->base + CRC_DR_OFFSET);
+
+		if (!WAIT_FOR((sys_read32(config->base + CRC_SR_OFFSET) & CRC_SR_DONE) != 0U,
+			      CRC_SF32LB_TIMEOUT_US, NULL)) {
+			ctx->state = CRC_STATE_IDLE;
+			crc_sf32lb_unlock(dev);
+			return -ETIMEDOUT;
+		}
+	}
+
+	/* Now we'll handle data that isn't 4-byte aligned */
+	if (idx < bufsize) {
+		uint32_t rem = 0U;
+		size_t rem_bytes = bufsize - idx;
+
+		for (size_t offset = 0U; offset < rem_bytes; offset++) {
+			rem |= (uint32_t)data_buf[idx + offset] << (offset * 8U);
+		}
+		idx += rem_bytes;
```

Comment:
You can make it shorter
```suggestion
		sys_get_le(&rem, &data_buf[idx + offset], rem_bytes);
```

Reply:
done